### PR TITLE
Assertion in scheduler.cpp with work_stealing algo and timed-out wait ops

### PR DESCRIPTION
**Pull request by saschazelzer**
I run into crashes in my application when using worker threads with the work_stealing scheduling algorithm and waits timing out on a buffered_channel.

It seems to be closely related to issue #166, but for a work_stealing scheduling algorithm. If requested, I could try and extract a minimal working example to reproduce this.

### CODE REVIEW

**Comment by olk** [comment]
please provide a minimal example - a fiber-context must not detached/attached from/to a scheduler while waiting for some event (stored in schedulers wait-list)

**Comment by saschazelzer** [comment]
All right, I will. To make sure there is no misunderstanding, the assertion is triggered when the wait operation timed out and the fiber is ready to be resumed.

**Comment by saschazelzer** [comment]
The following gist contains a self-contained example triggering the assertion:

```cpp
#include <iostream>
#include <boost/fiber/all.hpp>

int main() {

	boost::fibers::barrier barrier{ 2 };

	boost::fibers::mutex m;
	boost::fibers::condition_variable cv;

	boost::fibers::buffered_channel<int> channel{ 2 };

	int count = 2;

	std::thread([&] {
		boost::fibers::use_scheduling_algorithm<boost::fibers::algo::work_stealing>(2);
		barrier.wait();

		std::unique_lock<boost::fibers::mutex> l(m);
		cv.wait(l, [&] { return count == 0; });
	}).detach();

	boost::fibers::use_scheduling_algorithm<boost::fibers::algo::work_stealing>(2);
	barrier.wait();

	boost::fibers::fiber([&] {
		for (;;) {
			int x;
			if (channel.pop_wait_for(x, std::chrono::milliseconds(100)) ==
				boost::fibers::channel_op_status::timeout) {
				std::cout << "timed out" << std::endl;
			}
			else {
				std::cout << "got value: " << x << std::endl;
			}
		}
		count--;
	}).detach();

	boost::fibers::fiber([&] {
		for (int x = 0;; x++) {
			channel.push(x);
			boost::this_fiber::sleep_for(std::chrono::seconds(1));
		}
		count--;
	}).detach();

	std::unique_lock<boost::fibers::mutex> l(m);
	cv.wait(l, [&] { return count == 0; });

	return 0;
}
```

Tested on Windows 10, using a static 64-bit debug version of Boost 1.70. The program crashes with

    Assertion failed: ! ctx->wait_is_linked(), file ..\src\scheduler.cpp, line 405

**Comment by olk** [comment]
I don't use Windows as dev platform - on Linux the code works as expected.

**Comment by gpascualg** [comment]
I confirm being able to reproduce this with boost 1.73 on MSVC Version 16.7.0 Preview 1.0, building with /std:c++latest and /permissive- and linking against a debug library.

**Comment by olk** [comment]
should be fixed by #261
